docs/sdk/instructlab-sdk.md

```diff
@@ -0,0 +1,109 @@
+# InstructLab SDK
+
+## Motivation
+
+Today, the only way to "drive" the InstructLab opinionated workflow is via the `ilab` CLI. While this process provides a succinct way for everyday users to initialize a config, generate synthetic data, train a model, and evaluate it: the guardrails are quite limiting both in what a user can do and what the development team can add as features exposed directly to the user over time.
+
+Additionally, current consumers of InstructLab are finding themselves importing different library private and public APIs in combination with CLI functionality to achieve the workflows they want. While these more advanced usage patterns are not for everyone, providing ways to run bespoke and piecemeal workflows in a standardized and safe way for the community is a necessity.
+
+Unifying these various ranges of advanced workflows under an overarching `InstructLab SDK API` will allow for new usage patterns and a clearer story on what InstructLab can and should provide as user accessible endpoints.
+
+While each library can and _should_ have their own publicly accessible API, not all functionality being added to SDG, Training, and Eval needs to be correlated directly to the "InstructLab workflow". This Python SDK should, as the CLI does, expose an opinionated flow that uses functionality from the various libraries. The InstructLab SDK should be derived from the library APIs, not the other way around. SDG, for example, currently has a `generate_data` method, meant to only be accessed by InstructLab. This method simply calls other publicly available SDG functionality. Orchestration of the InstructLab flow like this should not be of concern to the individual libraries and instead be handled by the overarching InstructLab SDK which will maintain the user contracts. The InstructLab SDK will need to work within the bounds of what the Libraries expose as public APIs.
```

Comment:
(Thinking while reading) "an opinionated" (but perhaps LESS opinionated?) flow.

Reply:
> within the bounds of what the Libraries expose as public APIs

What's the current mechanism in which libraries communicate which APIs are public and which are not? Or is the definition of a "public API" here, loosely, "whatever ilab already pulls from the library"? Or, alternatively, that will be the result of "contract negotiation" that you mention below?

Reply:
most libraries have made APIs over time and are just now standardizing what they want public vs private. Part of the process of building this SDK will be making final decisions on that front.

In regard to opinionated, yes this will be less opinionated and more user driven

Reply:
I think the CLI will always be more "opinionated" than the SDK - in terms of functionality, we're exposing the same functionality to both "consumers" but the nature of the different interfaces will likely always allow for more flexibility with the SDK (since it's easier to orchestrate Python classes than CLI commands)